src/accumulator/proof.rs

```diff
@@ -0,0 +1,313 @@
+    ///   // 00   01  02   03  04   05  06   07
+    ///   // For proving 0, we need 01, 09 and 13's hashes. 00, 08, 12 and 14 can be calculated
+    ///   proof_hashes.push(bitcoin_hashes::sha256::Hash::from_str("4bf5122f344554c53bde2ebb8cd2b7e3d1600ad631c385a5d7cce23c7785459a").unwrap());
+    ///   proof_hashes.push(bitcoin_hashes::sha256::Hash::from_str("9576f4ade6e9bc3a6458b506ce3e4e890df29cb14cb5d3d887672aef55647a2b").unwrap());
+    ///   proof_hashes.push(bitcoin_hashes::sha256::Hash::from_str("29590a14c1b09384b94a2c0e94bf821ca75b62eacebc47893397ca88e3bbcbd7").unwrap());
+    ///   
+    ///   let mut hashes = Vec::new();
+    ///   for i in test_values {
+    ///     let mut engine = bitcoin_hashes::sha256::Hash::engine();
+    ///      engine.input(&[i]);
+    ///      let hash = sha256::Hash::from_engine(engine);
+    ///     hashes.push(hash);
+    ///   }
+    ///
+    ///   s.modify(&hashes, &vec![]);
+    ///   let p = Proof::new(targets, proof_hashes);
+    ///   assert!(p.verify(&vec![hashes[0]] , &s).expect("No error should happens"));
+    ///```
+    pub fn verify(&self, del_hashes: &Vec<sha256::Hash>, stump: &Stump) -> Result<bool, String> {
+        if self.targets.len() == 0 {
+            return Ok(true);
+        }
+
+        let mut calculated_roots = self
+            .create_root_candidates(del_hashes, stump)?
+            .into_iter()
+            .peekable();
+
+        let mut number_matched_roots = 0;
+        let mut roots_iter = stump.roots.iter().rev();
+
+        while let Some(root) = roots_iter.next() {
+            if let Some(next_calculated_root) = calculated_roots.peek() {
+                if *next_calculated_root == *root {
+                    number_matched_roots += 1;
+                    calculated_roots.next();
+                }
+            }
+        }
+
+        if calculated_roots.len() != number_matched_roots && calculated_roots.len() != 0 {
+            return Ok(false);
+        }
+        Ok(true)
+    }
+
+    pub fn create_root_candidates(
+        &self,
+        del_hashes: &Vec<sha256::Hash>,
+        stump: &Stump,
+    ) -> Result<Vec<sha256::Hash>, String> {
+        // Where all the root hashes that we've calculated will go to.
+        let total_rows = util::tree_rows(stump.leafs);
+
+        // Where all the parent hashes we've calculated in a given row will go to.
+        let mut calculated_root_hashes =
+            Vec::<sha256::Hash>::with_capacity(util::num_roots(stump.leafs) as usize);
+
+        // As we calculate nodes upwards, it accumulates here
+        let mut nodes: Vec<_> = self
+            .targets
+            .to_owned()
+            .into_iter()
+            .zip(del_hashes.to_owned())
+            .collect();
+
+        // Nodes must be sorted for finding siblings during hashing
+        nodes.sort();
+
+        // An iterator over proof hashes
+        let mut hashes_iter = self.hashes.iter();
+
+        for row in 0..(total_rows + 1) {
```

Comment:
I do think `for row in 0..=total_rows` is more readable.

Was this done for backwards compatibility reasons for rust version < 1.26? `rust-bitcoin`'s minimum version is 1.41.1 so the backwards compatibility problem wouldn't apply here.